src/hypothesis/extra/numpy.py

```diff
@@ -53,19 +53,24 @@ def from_dtype(dtype):
     elif dtype.kind == u'c':
         result = st.complex_numbers()
     elif dtype.kind in (u'S', u'a'):
-        result = st.binary()
+        result = st.binary(max_size=dtype.itemsize or None)
     elif dtype.kind == u'u':
-        result = st.integers(
-            min_value=0, max_value=1 << (4 * dtype.itemsize) - 1)
+        result = st.integers(min_value=0,
+                             max_value=2 ** (8 * dtype.itemsize) - 1)
     elif dtype.kind == u'i':
-        min_integer = -1 << (4 * dtype.itemsize - 1)
-        result = st.integers(min_value=min_integer, max_value=-min_integer - 1)
+        overflow = 2 ** (8 * dtype.itemsize - 1)
+        result = st.integers(min_value=-overflow, max_value=overflow - 1)
     elif dtype.kind == u'U':
-        result = st.text()
+        if dtype.itemsize == 0:  # pointer to arbitrary-length string
```

Comment:
Is this really represented as an itemsize of zero? That seems very odd. I'd expect it to be a pointer size.

Reply:
Misleading comment sorry - this dtype is *not* actually a pointer, it's a fixed but as yet undetermined length.  Will update the comment to  match.

Fun fact: you can also create dtypes for negative-length strings, which raise a MemoryError when you try to instantiate an array of them.  (I do not propose to check for this)